File: Simulation/Bronze/P1_2018-Team_Tic_Tac_Toe/tttt.py
ROWS = COLS = 3

board = open("tttt.in").read().splitlines()

def check_horizontal(teams):
    for r in range(ROWS):
        valid_char_count = 0
        curr_teams = set()
        for c in range(COLS):
            if board[r][c] in teams:
                valid_char_count += 1
                curr_teams.add(board[r][c])

        if valid_char_count == 3 and len(curr_teams) == len(teams):
            return True

    return False

def check_vertical(teams):
    for c in range(COLS):
        valid_char_count = 0
        curr_teams = set()
        for r in range(ROWS):
            if board[r][c] in teams:
                valid_char_count += 1
                curr_teams.add(board[r][c])

        if valid_char_count == 3 and len(curr_teams) == len(teams):
            return True
    return False

def check_diagonal(teams):
    # Main diagonal
    valid_char_count = 0
    curr_teams = set()
    for i in range(3):
        if board[i][i] in teams:
            valid_char_count += 1
            curr_teams.add(board[i][i])

    if valid_char_count == 3 and len(curr_teams) == len(teams):
        return True

    # Other diagonal
    valid_char_count = 0
    curr_teams = set()
    for i in range(3):
        if board[i][2-i] in teams:
            valid_char_count += 1
            curr_teams.add(board[i][2-i])

    return valid_char_count == 3 and len(curr_teams) == len(teams)

def check_win(teams):
    teams = set(teams)
    return (
        check_horizontal(teams) or
        check_vertical(teams) or
        check_diagonal(teams)
    )

cow_teams = set(board[r][c] for r in range(ROWS) for c in range(COLS))

single_wins = 0
for team in cow_teams:
    if check_win([team]):
        single_wins += 1

pair_wins = 0
for team1 in cow_teams:
    for team2 in cow_teams:
        if team1 < team2 and check_win([team1, team2]):
            pair_wins += 1

with open("tttt.out", "w") as fout:
    fout.write(f"{single_wins}\n")
    fout.write(f"{pair_wins}\n")

File: Simulation/Bronze/P1_2018-Team_Tic_Tac_Toe/tttt.cpp
#include <bits/stdc++.h>
#define ROWS 3
#define COLS 3

using namespace std;

string board[ROWS];

bool check_horizontal(unordered_set<char>& teams) {
    for (int r = 0; r < ROWS; r++) {
        int validCharCount = 0;
        unordered_set<char> currCowTeammates;
        for (int c = 0; c < COLS; c++) {
            if (teams.find(board[r][c]) != teams.end()) {
                currCowTeammates.insert(board[r][c]);
                validCharCount++;
            }
        }

        if (validCharCount == 3 && currCowTeammates.size() == teams.size())
            return true;
    }
    return false;
}

bool check_vertical(unordered_set<char>& teams) {
    for (int c = 0; c < COLS; c++) {
        int validCharCount = 0;
        unordered_set<char> currCowTeammates;
        for (int r = 0; r < ROWS; r++) {
            if (teams.find(board[r][c]) != teams.end()) {
                currCowTeammates.insert(board[r][c]);
                validCharCount++;
            }
        }

        if (validCharCount == 3 && currCowTeammates.size() == teams.size())
            return true;
    }
    return false;
}

bool check_diagonals(unordered_set<char>& teams) {
    int validCharCount = 0;
    unordered_set<char> currCowTeammates;
    for (int i = 0; i < ROWS; i++) {
        if (teams.find(board[i][i]) != teams.end()) {
            currCowTeammates.insert(board[i][i]);
            validCharCount++;
        }
    }

    if (validCharCount == 3 && currCowTeammates.size() == teams.size())
        return true;

    validCharCount = 0;
    currCowTeammates.clear();
    for (int i = 0; i < ROWS; i++) {
        if (teams.find(board[i][ROWS - i - 1]) != teams.end()) {
            currCowTeammates.insert(board[i][ROWS - i - 1]);
            validCharCount++;
        }
    }

    return validCharCount == 3 && currCowTeammates.size() == teams.size();
}

bool check_win(vector<char> teams) {
    unordered_set<char> ts(teams.begin(), teams.end());
    return check_horizontal(ts) ||
        check_vertical(ts) ||
        check_diagonals(ts);
}

int main() {
    freopen("tttt.in", "r", stdin);
    for (int r = 0; r < ROWS; r++)
        cin >> board[r];
    
    unordered_set<char> cowTeams;
    for (int r = 0; r < ROWS; r++)
        for (int c = 0; c < COLS; c++)
            cowTeams.insert(board[r][c]);

    int singleWins = 0;
    for (char ct : cowTeams)
        if (check_win({ct}))
            singleWins++;

    int doubleWins = 0;
    for (char ct1 : cowTeams)
        for (char ct2 : cowTeams)
            if (ct1 < ct2 && check_win({ct1, ct2}))
                doubleWins++;

    freopen("tttt.out", "w", stdout);
    cout << singleWins << endl;
    cout << doubleWins << endl;
    return 0;
}
